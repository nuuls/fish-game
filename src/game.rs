use js_sys::Math::random;

use crate::{
    level::Level,
    log,
    player::Player,
    types::{Entity, ShaderId, Triangle},
    user_input::{self, InputHandler},
};

pub struct Game {
    render_buffer: Vec<Triangle>,
    entities: Vec<Box<dyn Entity>>,
    input_handler: InputHandler,

    last_fps_print: f64,
    frames_drawn: usize,
}

struct ShitItem {
    id: String,
    triangles: Vec<Triangle>,
    moving: [f32; 9],
}

impl Entity for ShitItem {
    fn id(&self) -> &String {
        &self.id
    }
    fn triangles(&self) -> &Vec<crate::types::Triangle> {
        &self.triangles
    }
    fn update(&mut self, _time_passed: f32) {
        self.update();
    }
}

impl ShitItem {
    fn update(&mut self) {
        let tri = &mut self.triangles[0].coords;
        for n in 0..9 {
            let bounds = match n % 3 {
                0 => (23.0, 46.0),
                1 => (5.0, 10.0),
                _ => (0.0, 0.0),
            };
            let hit_edge = tri[n] <= bounds.0 || tri[n] >= bounds.1;

            let movement = match hit_edge {
                true => -self.moving[n],
                false => self.moving[n],
            };

            tri[n] = tri[n] + movement;
            self.moving[n] = movement;
        }
    }
}

impl Game {
    pub fn new() -> Game {
        let mut entities: Vec<Box<dyn Entity>> = random_shit_items(3)
            .into_iter()
            .map(|si| Box::new(si) as Box<dyn Entity>)
            .collect();

        let level = Level::load_from_svg_str(include_str!("../assets/map.svg"));
        let player = Box::new(Player::new(level.player_pos()));
        entities.push(Box::new(level));
        entities.push(player);

        let mut input_handler = user_input::InputHandler::new();
        input_handler.attach();

        Game {
            render_buffer: vec![],
            entities,
            input_handler,
            frames_drawn: 0,
            last_fps_print: 0.0,
        }
    }

    pub fn next_frame(&mut self, time_passed: f32) -> &Vec<Box<dyn Entity>> {
        self.render_buffer.clear();
        // data
        let input = self.input_handler.current_state();
        for item in &mut self.entities {
            item.on_user_input(&input);
            item.update(time_passed);
        }

        self.handle_fps();

        &self.entities
    }

    fn handle_fps(&mut self) {
        self.frames_drawn += 1;

        if self.frames_drawn % 100 == 0 {
            let window = web_sys::window().expect("should have a window in this context");
            let performance = window
                .performance()
                .expect("performance should be available");
            let now = performance.now();
            let fps = 100.0 * (1.0 / ((now - self.last_fps_print) / 1000.0));
            log!("FPS {:.2}", fps);
            self.last_fps_print = now;
        }
    }
}

fn random_shit_items(n: usize) -> Vec<ShitItem> {
    (0..n)
        .map(|_| {
            let mut t = [0.0; 9];
            let mut moving = [0.0; 9];
            for i in 0..3 {
                t[i * 3] = (random() * 23.0 + 23.0) as f32;
                t[i * 3 + 1] = (random() * 5.0 + 5.0) as f32;

                moving[i * 3] = (random() * 0.01 - 0.01) as f32;
                moving[i * 3 + 1] = (random() * 0.01 - 0.01) as f32;
            }
            ShitItem {
                id: format!("shit_item-{}", n),
                triangles: vec![Triangle {
                    coords: t,
                    color: [0.9; 4],
                    shader_id: ShaderId::Default,
                }],
                moving,
            }
        })
        .collect()
}
