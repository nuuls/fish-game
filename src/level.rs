use crate::game::Triangle;
use crate::log;
use regex::Regex;
use svg;
use svg::node::element::path::{Command, Data, Position};
use svg::node::element::Style;
use svg::parser::Event;

pub struct Level {
    triangles: Vec<Triangle>,
}

fn update_point(
    (x, y): &mut (f32, f32),
    position: Position,
    (new_x, new_y): (Option<f32>, Option<f32>),
) -> (f32, f32) {
    match position {
        Position::Absolute => {
            *x = new_x.unwrap_or(*x);
            *y = new_y.unwrap_or(*y);
        }
        Position::Relative => {
            *x += new_x.unwrap_or(0.0);
            *y += new_y.unwrap_or(0.0);
        }
    }
    (*x, *y)
}

impl Level {
    #[allow(dead_code)]
    pub fn load_from_svg_str(content: &str) -> Level {
        Self::parse(svg::read(content).unwrap())
    }

    fn parse(parser: svg::parser::Parser) -> Level {
        let mut points: Vec<Vec<f32>> = vec![vec![0.0, 0.0]];
        let mut color = [1.0, 0.0, 1.0, 1.0];

        for event in parser {
            match event {
                Event::Tag("path", _, attributes) => {
                    let data = attributes.get("d").unwrap();
                    let data = Data::parse(data).unwrap();

                    color = color_from_style(
                        attributes.get("style").map(|v| Style::new(v.to_string())),
                    )
                    .unwrap_or(color);

                    let mut current_pos = (0.0, 0.0);

                    for command in data.iter() {
                        match command {
                            Command::Move(position, parameters) => {
                                let start = parameters
                                    .array_chunks::<2>()
                                    .last()
                                    .map(|[x, y]| {
                                        update_point(
                                            &mut current_pos,
                                            *position,
                                            (Some(*x), Some(*y)),
                                        )
                                    })
                                    .unwrap_or((0.0, 0.0));

                                points.push(vec![start.0, start.1]);
                            }
                            Command::Line(position, parameters) => {
                                parameters.array_chunks::<2>().for_each(|[x, y]| {
                                    if let Some(path) = points.last_mut() {
                                        let p = update_point(
                                            &mut current_pos,
                                            *position,
                                            (Some(*x), Some(*y)),
                                        );

                                        path.push(p.0);
                                        path.push(p.1);
                                    };
                                });
                            }
                            Command::HorizontalLine(position, parameters) => {
                                parameters.iter().for_each(|x| {
                                    if let Some(path) = points.last_mut() {
                                        let p = update_point(
                                            &mut current_pos,
                                            *position,
                                            (Some(*x), None),
                                        );

                                        path.push(p.0);
                                        path.push(p.1);
                                    };
                                });
                            }
                            Command::VerticalLine(position, parameters) => {
                                parameters.iter().for_each(|y| {
                                    if let Some(path) = points.last_mut() {
                                        let p = update_point(
                                            &mut current_pos,
                                            *position,
                                            (None, Some(*y)),
                                        );

                                        path.push(p.0);
                                        path.push(p.1);
                                    };
                                });
                            }
                            Command::Close => {
                                points.push(vec![0.0, 0.0]);
                            }
                            _ => {
                                log!("unsupported svg path command: {:?}", command);
                            }
                        }
                    }
                }
                Event::Tag("rect", _, attributes) => {
                    let x = attributes.get("x").unwrap().parse::<f32>().unwrap();
                    let y = attributes.get("y").unwrap().parse::<f32>().unwrap();
                    let width = attributes.get("width").unwrap().parse::<f32>().unwrap();
                    let height = attributes.get("height").unwrap().parse::<f32>().unwrap();

                    color = color_from_style(
                        attributes.get("style").map(|v| Style::new(v.to_string())),
                    )
                    .unwrap_or(color);

                    let path = vec![
                        // first triangle
                        x,
                        y,
                        x + width,
                        y,
                        x + width,
                        y + height,
                        // second triangle
                        x,
                        y,
                        x + width,
                        y + height,
                        x,
                        y + height,
                    ];

                    points.push(path);
                }
                _ => {}
            }
        }

        let mut triangles: Vec<Triangle> = vec![];

        for path in points.iter() {
            // not even a triangle
            if path.len() < 6 {
                continue;
            }

            let path_triangles = earcutr::earcut(path.as_slice(), &vec![], 2);

            for [a, b, c] in path_triangles.array_chunks::<3>() {
                triangles.push(Triangle {
                    coords: [
                        path[*a * 2] as f32,
                        path[*a * 2 + 1] as f32,
                        0.0 as f32,
                        path[*b * 2] as f32,
                        path[*b * 2 + 1] as f32,
                        0.0 as f32,
                        path[*c * 2] as f32,
                        path[*c * 2 + 1] as f32,
                        0.0 as f32,
                    ],
                    color,
                });
            }
        }

        Level { triangles }
    }

    pub fn triangles(&self) -> &Vec<Triangle> {
        &self.triangles
    }
}

fn color_from_style(style: Option<Style>) -> Option<[f32; 4]> {
    let style = style?;
    let style_str = style.to_string();

    let re = Regex::new(r"fill:#([0-9a-z]{6})").unwrap();

    let rgb_str = re.captures(&style_str)?.get(1)?.as_str();

    let rgb = hex::decode(&rgb_str.as_bytes()).unwrap();
    let vals: Vec<f32> = rgb.into_iter().map(|v| v as f32 / 255.0).collect();

    let mut out = [1.0; 4];
    for i in 0..3 {
        out[i] = vals[i];
    }
    Some(out)
}

#[cfg(test)]
mod tests {
    #[test]
    fn example() {
        let content = r#"<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0 0 300 300" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" xmlns:serif="http://www.serif.com/" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:1.5;">
    <path d="M49,144L141,146L152,173L172,187L207,191L241,186L255,165L259,147L292,147L289,268L42,274L49,144Z" style="fill:none;stroke:black;stroke-width:1px;"/>
</svg>"#;

        let level = super::Level::load_from_svg_str(content);

        insta::assert_debug_snapshot!(&level
            .triangles
            .iter()
            .map(|t| t.coords)
            .collect::<Vec<_>>());
    }
}
