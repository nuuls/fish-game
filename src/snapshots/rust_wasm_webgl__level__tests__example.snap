---
source: src/level.rs
expression: "&level.triangles"
---
[
    [
        42.0,
        274.0,
        0.0,
        49.0,
        144.0,
        0.0,
        141.0,
        146.0,
        0.0,
    ],
    [
        255.0,
        165.0,
        0.0,
        259.0,
        147.0,
        0.0,
        292.0,
        147.0,
        0.0,
    ],
    [
        42.0,
        274.0,
        0.0,
        141.0,
        146.0,
        0.0,
        152.0,
        173.0,
        0.0,
    ],
    [
        241.0,
        186.0,
        0.0,
        255.0,
        165.0,
        0.0,
        292.0,
        147.0,
        0.0,
    ],
    [
        42.0,
        274.0,
        0.0,
        152.0,
        173.0,
        0.0,
        172.0,
        187.0,
        0.0,
    ],
    [
        241.0,
        186.0,
        0.0,
        292.0,
        147.0,
        0.0,
        289.0,
        268.0,
        0.0,
    ],
    [
        289.0,
        268.0,
        0.0,
        42.0,
        274.0,
        0.0,
        172.0,
        187.0,
        0.0,
    ],
    [
        207.0,
        191.0,
        0.0,
        241.0,
        186.0,
        0.0,
        289.0,
        268.0,
        0.0,
    ],
    [
        289.0,
        268.0,
        0.0,
        172.0,
        187.0,
        0.0,
        207.0,
        191.0,
        0.0,
    ],
]
